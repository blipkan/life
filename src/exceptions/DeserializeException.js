export class DeserializeException extends Error {
  constructor({ className, sourceString, causeError, message }) {
    const errorMessage =
      message ||
      `cannot deserialize ${className} from string '${sourceString}'. Due to: \n ${causeError}`

    super(errorMessage)
    // TODO: check stack
    this.stack = new Error().stack
    this.name = this.constructor.name
  }
}
