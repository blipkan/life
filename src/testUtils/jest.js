/* eslint-env jest */
import last from 'lodash/last'
import join from 'lodash/join'
import noop from 'lodash/noop'

export const mockMethod = (
  object,
  methodName,
  customImplementation,
  accessType
) => {
  let method
  const getMethod = () => method

  const getCalls = () => method.mock.calls

  const getCall = (callNumber = -1) => {
    const calls = getCalls()
    return callNumber === -1 ? last(calls) : calls[callNumber]
  }

  beforeAll(() => {
    method = jest.spyOn(object, methodName, accessType)
    if (customImplementation) {
      method.mockImplementation(customImplementation)
    }
  })

  afterEach(() => {
    method.mockClear()
  })

  afterAll(() => {
    method.mockReset()
    method.mockRestore()
  })

  return {
    method,
    getMethod,
    getCall,
    getCalls
  }
}

export const mockConsoleMethod = (
  consoleMethodName,
  customImplementation = noop
) => {
  const DEFAULT_METHOD_NAME = 'error'
  const methodName = ['error', 'warn', 'log', 'info'].includes(
    consoleMethodName
  )
    ? consoleMethodName
    : DEFAULT_METHOD_NAME

  const { method, getMethod, getCall, getCalls } = mockMethod(
    global.console,
    methodName,
    customImplementation
  )

  const getConsoleMessage = callNumber => {
    return join(getCall(callNumber), ' ')
  }

  return { getConsoleMessage, getMethod, method, getCalls }
}

export const spyConsoleMethod = consoleMethodName =>
  mockConsoleMethod(consoleMethodName, null)
