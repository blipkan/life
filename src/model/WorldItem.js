/* eslint-disable no-debugger */
import find from 'lodash/find'
import values from 'lodash/values'
import split from 'lodash/split'
import isEmpty from 'lodash/isEmpty'
import map from 'lodash/map'

import { checkKey, consoleMessage, makeId } from '../utils/misc'
import { DeserializeException } from '../exceptions/DeserializeException'

const WorldItemTypeKey = {
  DEAD: 'DEAD',
  LIVE: 'LIVE',
  LIFE: 'LIFE'
}

const WorldItemStatusKey = {
  DELETED: 'DELETED',
  NEW: 'NEW',
  PROCESSED: 'PROCESSED'
}

const WorldItemStatus = {
  [WorldItemStatusKey.NEW]: {
    key: WorldItemStatusKey.NEW,
    intValue: 0
  },
  [WorldItemStatusKey.DELETED]: {
    key: WorldItemStatusKey.DELETED,
    intValue: 1
  },
  [WorldItemStatusKey.PROCESSED]: {
    key: WorldItemStatusKey.PROCESSED,
    intValue: 2
  }
}

const WorldItemType = {
  [WorldItemTypeKey.LIVE]: {
    key: WorldItemTypeKey.LIVE,
    iconKey: 'live',
    continuousCount: 2,
    displayOrder: 1,
    intValue: 1
  },

  [WorldItemTypeKey.DEAD]: {
    key: WorldItemTypeKey.DEAD,
    iconKey: 'dead',
    continuousCount: 3,
    displayOrder: 2,
    intValue: 2
  },

  [WorldItemTypeKey.LIFE]: {
    key: WorldItemTypeKey.LIFE,
    iconKey: 'birth',
    continuousCount: 0,
    displayOrder: 0,
    intValue: 3
  }
}

class WorldItem {
  #id
  #statusKey
  #type

  static SERIALIZE_DELIMITER = ','

  constructor(typeKey, statusKey = WorldItem.Status.NEW.key) {
    this.checkConstructorArgs(typeKey, statusKey)

    this.#id = makeId()
    this.#statusKey = statusKey
    this.#type = WorldItemType[typeKey]
  }

  checkConstructorArgs(typeKey, statusKey) {
    const invalidArgs = this.constructor.validateConstructorArgs(
      typeKey,
      statusKey
    )

    if (!isEmpty(invalidArgs)) {
      throw new Error(
        `Cannot create ${this.constructor.name} by arguments: ${map(
          invalidArgs,
          arg => `${arg.name}:${arg.value}`
        )}`
      )
    }
  }

  get id() {
    return this.#id
  }

  get iconKey() {
    return this.#type.iconKey
  }

  get continuousCount() {
    return this.#type.continuousCount
  }

  get typeKey() {
    return this.#type.key
  }

  get statusKey() {
    return this.#statusKey
  }

  set statusKey(statusKey) {
    this.#statusKey = statusKey || WorldItem.Status.NEW.key
  }

  get statusIntValue() {
    return WorldItemStatus[this.statusKey].intValue
  }

  clone() {
    return WorldItem.create(this.typeKey, this.statusKey)
  }

  update(valuesMap = {}) {
    Object.entries(valuesMap).forEach(([key, value]) => {
      this[key] = value
    })
    return this
  }

  serialize() {
    const typeValue = this.#type.intValue
    const statusValue = this.statusIntValue

    return statusValue === WorldItem.Status.NEW.intValue
      ? typeValue
      : [typeValue, statusValue].join(WorldItem.SERIALIZE_DELIMITER)
  }
}

WorldItem.deserialize = serialized => {
  try {
    const storedValues = split(serialized, WorldItem.SERIALIZE_DELIMITER)
    const typeValue = storedValues[0] ? +storedValues[0] : undefined
    const statusValue = +storedValues[1]

    const typeKey = find(values(WorldItemType), { intValue: typeValue })?.key
    const statusKey = find(values(WorldItemStatus), { intValue: statusValue })
      ?.key

    return new WorldItem(typeKey, statusKey)
  } catch (e) {
    throw new DeserializeException({
      className: WorldItem.prototype.constructor.name,
      sourceString: serialized,
      causeError: e
    })
  }
}

WorldItem.isTypeKeyValid = typeKey => checkKey(WorldItem.Type, typeKey)
WorldItem.isStatusKeyValid = statusKey => checkKey(WorldItem.Status, statusKey)

WorldItem.validateConstructorArgs = (typeKey, statusKey) => {
  const invalidArgs = []

  if (!WorldItem.isTypeKeyValid(typeKey)) {
    invalidArgs.push({ name: 'typeKey', value: typeKey })
  }

  if (!WorldItem.isStatusKeyValid(statusKey)) {
    invalidArgs.push({ name: 'statusKey', value: statusKey })
  }

  return invalidArgs
}

WorldItem.create = (typeKey, statusKey = WorldItem.Status.NEW.key) => {
  try {
    return new WorldItem(typeKey, statusKey)
  } catch (e) {
    consoleMessage(e.message)
  }
}

WorldItem.Type = WorldItemType
WorldItem.Status = WorldItemStatus

export default WorldItem
