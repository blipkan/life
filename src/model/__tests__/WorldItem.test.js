/* eslint-env jest */

import isEqual from 'lodash/isEqual'

import WorldItem from '../WorldItem'
import { mockConsoleMethod, spyConsoleMethod } from '../../testUtils/jest'
import { DeserializeException } from '../../exceptions/DeserializeException'

const ParamName = {
  typeKey: 'typeKey',
  statusKey: 'statusKey'
}

const TestType = WorldItem.Type.LIFE
const TestStatus = WorldItem.Status.PROCESSED
const callConstructor = (...args) => new WorldItem(...args)
const callCreate = (...args) => WorldItem.create(...args)
const createInstance = () => callCreate(TestType.key)

describe('WorldItem.constructor SUCCESS', () => {
  it('constructor - uniq id', () => {
    const instance = callConstructor(TestType.key)
    const instance2 = callConstructor(TestType.key)
    expect(instance.id).not.toBe(instance2.id)
  })

  it('constructor - valid typeKey', () => {
    const instance = callConstructor(TestType.key)
    expect(instance.typeKey).toBe(TestType.key)
    expect(instance.iconKey).toBe(TestType.iconKey)
    expect(instance.continuousCount).toBe(TestType.continuousCount)
  })

  it('constructor - valid statusKey', () => {
    const instance = callConstructor(TestType.key, TestStatus.key)
    expect(instance.statusKey).toBe(TestStatus.key)
    expect(instance.statusIntValue).toBe(TestStatus.intValue)
  })

  it('constructor - undefined statusKey (default=NEW)', () => {
    const defaultStatus = WorldItem.Status.NEW
    const instance = callConstructor(TestType.key)
    expect(instance.statusKey).toBe(defaultStatus.key)
    expect(instance.statusIntValue).toBe(defaultStatus.intValue)
  })
})

describe('WorldItem.constructor WRONG', () => {
  mockConsoleMethod()

  it('constructor - undefined typeKey (throw Error)', () => {
    const expectation = expect(() => callConstructor())
    expectation.toThrowError(ParamName.typeKey)
    expectation.toThrowError('undefined')
    expectation.toThrowErrorMatchingSnapshot()
  })

  it('constructor - wrong typeKey (throw Error)', () => {
    const wrongTypeKey = 'wrongTypeKey'
    const expectation = expect(() => callConstructor(wrongTypeKey))
    expectation.toThrowError(ParamName.typeKey)
    expectation.toThrowError(wrongTypeKey)
    expectation.toThrowErrorMatchingSnapshot()
  })

  it('constructor - wrong statusKey (throw Error)', () => {
    const wrongStatusKey = 'wrongStatusKey'
    const expectation = expect(() =>
      callConstructor(TestType.key, wrongStatusKey)
    )
    expectation.toThrowError(ParamName.statusKey)
    expectation.toThrowError(wrongStatusKey)
    expectation.toThrowErrorMatchingSnapshot()
  })

  it('constructor - wrong typeKey AND wrong statusKey (throw Error)', () => {
    const wrongTypeKey = 'wrongTypeKey'
    const wrongStatusKey = 'wrongStatusKey'
    const expectation = expect(() =>
      callConstructor(wrongTypeKey, wrongStatusKey)
    )
    expectation.toThrowError(ParamName.typeKey)
    expectation.toThrowError(wrongTypeKey)
    expectation.toThrowError(ParamName.statusKey)
    expectation.toThrowError(wrongStatusKey)
    expectation.toThrowErrorMatchingSnapshot()
  })
})

describe('WorldItem.create SUCCESS', () => {
  it('create - uniq id', () => {
    const instance = callCreate(TestType.key)
    const instance2 = callCreate(TestType.key)
    expect(instance.id).not.toBe(instance2.id)
  })

  it('create - valid typeKey', () => {
    const instance = callCreate(TestType.key)
    expect(instance.typeKey).toBe(TestType.key)
    expect(instance.iconKey).toBe(TestType.iconKey)
    expect(instance.continuousCount).toBe(TestType.continuousCount)
  })

  it('create - valid statusKey', () => {
    const instance = callCreate(TestType.key, TestStatus.key)
    expect(instance.statusKey).toBe(TestStatus.key)
    expect(instance.statusIntValue).toBe(TestStatus.intValue)
  })

  it('create - undefined statusKey (default=NEW)', () => {
    const defaultStatus = WorldItem.Status.NEW
    const instance = callCreate(TestType.key)
    expect(instance.statusKey).toBe(defaultStatus.key)
    expect(instance.statusIntValue).toBe(defaultStatus.intValue)
  })
})

describe('WorldItem.create - WRONG', () => {
  const { getConsoleMessage, getMethod } = mockConsoleMethod()

  const performChecks = (paramName, paramValue, callTimes, instance) => {
    expect(instance).toBeUndefined()
    expect(getMethod()).toHaveBeenCalledTimes(callTimes)
    expect(getConsoleMessage(0)).toContain(paramValue)
    const message = getConsoleMessage()
    expect(message).toContain(paramName)
    expect(message).toContain(paramValue)
  }

  it('create - undefined typeKey (undefined instance, console.error)', () => {
    const instance = callCreate()
    performChecks(ParamName.typeKey, undefined, 2, instance)
  })

  it('create - wrong typeKey (undefined instance, console.error)', () => {
    const wrongTypeKey = 'wrongTypeKey'
    const instance = callCreate(wrongTypeKey)
    performChecks(ParamName.typeKey, wrongTypeKey, 2, instance)
  })

  it('create - wrong statusKey (undefined instance, console.error)', () => {
    const wrongStatusKey = 'wrongStatusKey'
    const instance = callCreate(TestType.key, wrongStatusKey)
    performChecks(ParamName.statusKey, wrongStatusKey, 2, instance)
  })

  it('create - wrong typeKey AND wrong statusKey (undefined instance, console.error)', () => {
    const wrongTypeKey = 'wrongTypeKey'
    const wrongStatusKey = 'wrongStatusKey'
    const instance = callCreate(wrongTypeKey, wrongStatusKey)
    performChecks(ParamName.typeKey, wrongTypeKey, 3, instance)
    expect(getConsoleMessage(1)).toContain(wrongStatusKey)
    const message = getConsoleMessage()
    expect(message).toContain(ParamName.statusKey)
    expect(message).toContain(wrongStatusKey)
  })
})

describe('WorldItem.clone SUCCESS', () => {
  it('clone', () => {
    const instance = createInstance()
    const cloned = instance.clone()
    expect(instance).not.toBe(cloned)
    expect(instance.id).not.toBe(cloned.id)
    expect(instance).toEqual(cloned)
    expect(isEqual(instance, cloned)).toBe(true)
  })
})

describe('WorldItem.update SUCCESS', () => {
  it('update', () => {
    const newStatusKey = WorldItem.Status.PROCESSED.key
    const instance = createInstance()
    instance.update({
      [ParamName.statusKey]: newStatusKey
    })
    expect(instance.statusKey).toBe(newStatusKey)
  })

  it('update readOnly - throw Error', () => {
    const newTypeKey = WorldItem.Type.DEAD.key
    const instance = createInstance()
    const expectation = expect(() =>
      instance.update({
        [ParamName.typeKey]: newTypeKey
      })
    )
    expectation.toThrowError()
    expectation.toThrowErrorMatchingSnapshot()
  })
})

describe('WorldItem.serialize', () => {
  const testSerialize = (type, status) => {
    const expectedSerialized =
      status === WorldItem.Status.NEW
        ? type.intValue
        : `${type.intValue}${WorldItem.SERIALIZE_DELIMITER}${status.intValue}`

    const instance = callCreate(type.key, status.key)
    const serialized = instance.serialize()
    expect(serialized).toBe(expectedSerialized)
  }

  describe('WorldItem.serialize [Type]', () => {
    const theStatus = TestStatus

    it('serialize type = LIVE', () => {
      const theType = WorldItem.Type.LIVE
      testSerialize(theType, theStatus)
    })

    it('serialize type = DEAD', () => {
      const theType = WorldItem.Type.DEAD
      testSerialize(theType, theStatus)
    })

    it('serialize type = LIFE', () => {
      const theType = WorldItem.Type.LIFE
      testSerialize(theType, theStatus)
    })
  })

  describe('WorldItem.serialize [Status]', () => {
    const theType = TestType

    it('serialize type = NEW', () => {
      const theStatus = WorldItem.Status.NEW
      testSerialize(theType, theStatus)
    })

    it('serialize type = PROCESSED', () => {
      const theStatus = WorldItem.Status.PROCESSED
      testSerialize(theType, theStatus)
    })

    it('serialize type = DELETED', () => {
      const theStatus = WorldItem.Status.DELETED
      testSerialize(theType, theStatus)
    })
  })
})

describe('WorldItem.deserialize', () => {
  const testDeserialize = (type, status) => {
    const serialized =
      status === WorldItem.Status.NEW
        ? `${type.intValue}`
        : `${type.intValue}${WorldItem.SERIALIZE_DELIMITER}${status.intValue}`

    const instance = WorldItem.deserialize(serialized)
    expect(instance).toBeDefined()
    expect(instance.typeKey).toBe(type.key)
    expect(instance.statusKey).toBe(status.key)
  }

  describe('WorldItem.deserialize [Type]', () => {
    const theStatus = TestStatus

    it('deserialize type = LIVE', () => {
      const theType = WorldItem.Type.LIVE
      testDeserialize(theType, theStatus)
    })

    it('deserialize type = DEAD', () => {
      const theType = WorldItem.Type.DEAD
      testDeserialize(theType, theStatus)
    })

    it('deserialize type = LIFE', () => {
      const theType = WorldItem.Type.LIFE
      testDeserialize(theType, theStatus)
    })
  })

  describe('WorldItem.deserialize [Status]', () => {
    const theType = TestType

    it('deserialize type = NEW', () => {
      const theStatus = WorldItem.Status.NEW
      testDeserialize(theType, theStatus)
    })

    it('deserialize type = PROCESSED', () => {
      const theStatus = WorldItem.Status.PROCESSED
      testDeserialize(theType, theStatus)
    })

    it('deserialize type = DELETED', () => {
      const theStatus = WorldItem.Status.DELETED
      testDeserialize(theType, theStatus)
    })
  })

  describe('WorldItem.deserialize [WRONG input string]', () => {
    mockConsoleMethod()

    const runTest = wrongString => {
      const expectation = expect(() => WorldItem.deserialize(wrongString))
      expectation.toThrowError(DeserializeException)
      expectation.toThrowError('WorldItem')
      expectation.toThrowError(`${wrongString}`)
    }

    it('deserialize from wrong string', () => {
      runTest('some unparsible string')
    })

    it('deserialize from empty string', () => {
      runTest('')
    })

    it('deserialize from null string', () => {
      runTest(null)
    })

    it('deserialize from undefined string', () => {
      runTest(undefined)
    })
  })
})
