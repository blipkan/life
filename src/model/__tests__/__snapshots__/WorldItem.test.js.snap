// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorldItem.constructor WRONG constructor - undefined typeKey (throw Error) 1`] = `"Cannot create WorldItem by arguments: typeKey:undefined"`;

exports[`WorldItem.constructor WRONG constructor - wrong statusKey (throw Error) 1`] = `"Cannot create WorldItem by arguments: statusKey:wrongStatusKey"`;

exports[`WorldItem.constructor WRONG constructor - wrong typeKey (throw Error) 1`] = `"Cannot create WorldItem by arguments: typeKey:wrongTypeKey"`;

exports[`WorldItem.constructor WRONG constructor - wrong typeKey AND wrong statusKey (throw Error) 1`] = `"Cannot create WorldItem by arguments: typeKey:wrongTypeKey,statusKey:wrongStatusKey"`;

exports[`WorldItem.update SUCCESS update readOnly - throw Error 1`] = `"Cannot set property typeKey of #<WorldItem> which has only a getter"`;
